source/base/utilities.cc

```diff
@@ -91,6 +90,169 @@ namespace Utilities
+    // https://gitlab.com/octopus-code/octopus/blob/develop/src/grid/hilbert.c
+    // https://github.com/trilinos/Trilinos/blob/master/packages/zoltan/src/hsfc/hsfc_hilbert.c
+    // (Zoltan_HSFC_InvHilbertXd)
+    // https://github.com/aditi137/Hilbert/blob/master/Hilbert/hilbert.cpp
+
+    // now we can map to 1D coordinate stored in Transpose format
+    // adopt AxestoTranspose function from the paper, that
+    // transforms in-place between geometrical axes and Hilbert transpose.
+    // Example:   b=5 bits for each of n=3 coordinates.
+    //            15-bit Hilbert integer = A B C D E F G H I J K L M N O is
+    //            stored as its Transpose
+    //                   X[0] = A D G J M                X[2]|
+    //                   X[1] = B E H K N    <------->       | /X[1]
+    //                   X[2] = C F I L O               axes |/
+    //                          high  low                    0------ X[0]
+
+    // Depth of the Hilbert curve
+    Assert(bit_per_dim <= std::numeric_limits<Integer>::digits,
+           ExcMessage("This integer type can not hold " +
+                      std::to_string(bit_per_dim) + " bits."));
+
+    const Integer M = 1 << (bit_per_dim - 1); // largest bit
+
+    for (unsigned int index = 0; index < int_points.size(); ++index)
+      {
+        auto &X = int_points[index];
+        auto &L = res[index];
+
+        // Inverse undo
+        for (Integer q = M; q > 1; q >>= 1)
+          {
+            const Integer p = q - 1;
+            for (unsigned int i = 0; i < dim; i++)
+              {
+                // invert
+                if (X[i] & q)
+                  {
+                    X[0] ^= p;
+                  }
+                // exchange
+                else
+                  {
+                    const Integer t = (X[0] ^ X[i]) & p;
+                    X[0] ^= t;
+                    X[i] ^= t;
+                  }
+              }
+          }
+
+        // Gray encode (inverse of decode)
+        for (unsigned int i = 1; i < dim; i++)
+          X[i] ^= X[i - 1];
+
+        Integer t = 0;
+        for (Integer q = M; q > 1; q >>= 1)
+          if (X[dim - 1] & q)
+            t ^= q - 1;
+        for (unsigned int i = 0; i < dim; i++)
+          X[i] ^= t;
+
+        // now we need to go from index stored in transpose format to
+        // consecutive format, which is better suited for comparators.
+        // we could inteleave into some big unsigned int...
+        // https://www.forceflow.be/2013/10/07/morton-encodingdecoding-through-bit-interleaving-implementations/
+        // https://stackoverflow.com/questions/4431522/given-2-16-bit-ints-can-i-interleave-those-bits-to-form-a-single-32-bit-int
+        // ...but we would loose spatial resolution!
+
+        // Inteleave using brute force, follow TransposetoLine from
+        // https://github.com/aditi137/Hilbert/blob/master/Hilbert/hilbert.cpp
```

Comment:
what's the license of that code?

Reply:
there is none in that repo. 

But de-facto it's not one-to-one copy of [author's code](https://github.com/aditi137/Hilbert/blob/master/Hilbert/hilbert.cpp#L120-L134) due to (i) C vs C++ (ii) typenames (iii) formatting.

Reply:
That's the wrong criterion. For the copyright it doesn't matter whether you indent it differently, but whether the intellectual contribution is the same. If you translate it 1:1 from C to C++, then that's not a significant intellectual contribution. If you translated it into assembler language or Forth, then that probably is.

I suspect that it doesn't matter her, though. The code is short enough so that every reasonable programmer would have written it this way.